<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [154]:
import pandas as pd
import psycopg2

In [156]:
#Создаем переменную запроса для подключения к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [157]:
query_3_1 = f'''select COUNT(*)
                from public.vacancies
            '''

In [158]:
answer_3_1 = pd.read_sql_query(query_3_1, connection)
answer_3_1

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\2262139106.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_1 = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [159]:
query_3_2 = f'''select COUNT(*)
                from public.employers
            '''

In [160]:
answer_3_2 = pd.read_sql_query(query_3_2, connection)
answer_3_2

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1178497560.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [161]:
query_3_3 = f'''select COUNT(*)
                from public.areas
            '''

In [162]:
answer_3_3 = pd.read_sql_query(query_3_3, connection)
answer_3_3

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\4098136936.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [163]:
query_3_4 = f'''select COUNT(*)
                from public.industries
            '''

In [164]:
answer_3_4 = pd.read_sql_query(query_3_4, connection)
answer_3_4

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\3272017797.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Из предварительного анализа получили следующие данные об используемой базе:
- Общее количество вакансий равно 49197 
- Общее количество работодателей равно 23501 
- Общее количество регионов равно 1362 
- Общее количество сфер деятельности равно 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [165]:
query_4_1 = f'''select COUNT(*),
                    a.name
                from public.vacancies v
                JOIN public.areas a ON a.id = v.area_id
                GROUP BY a.name
                ORDER BY 1 DESC
            '''

In [166]:
answer_4_1 = pd.read_sql_query(query_4_1, connection)
answer_4_1

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1030361834.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_1 = pd.read_sql_query(query_4_1, connection)


count             name
0     5333           Москва
1     2851  Санкт-Петербург
2     2112            Минск
3     2006      Новосибирск
4     1892           Алматы
..     ...              ...
764      1       Тарко-Сале
765      1    Новоаннинский
766      1            Бирск
767      1           Сасово
768      1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [167]:
query_4_2 = f'''select COUNT(*)
                from public.vacancies
                WHERE salary_from IS NOT NULL
                    OR salary_to IS NOT NULL
            '''

In [168]:
answer_4_2 = pd.read_sql_query(query_4_2, connection)
answer_4_2

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\4247165315.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [169]:
query_4_3 = f'''select ROUND(AVG(salary_from), 0) Нижняя_граница,
                    ROUND(AVG(salary_to), 0) Верхняя_граница
                from public.vacancies
                WHERE salary_from IS NOT NULL
                    OR salary_to IS NOT NULL
            '''

In [170]:
answer_4_3 = pd.read_sql_query(query_4_3, connection)
answer_4_3

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1371674522.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_3 = pd.read_sql_query(query_4_3, connection)


Нижняя_граница  Верхняя_граница
0         71065.0         110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [171]:
query_4_4 = f'''select COUNT(*),
                    schedule,
                    employment
                from public.vacancies
                GROUP BY schedule, employment
                ORDER BY 1 DESC
            '''

In [172]:
answer_4_4 = pd.read_sql_query(query_4_4, connection)
answer_4_4

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1828156850.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_4 = pd.read_sql_query(query_4_4, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [173]:
query_4_5 = f'''select COUNT(*),
                    experience
                from public.vacancies
                GROUP BY experience
                ORDER BY 1 ASC
            '''

In [174]:
answer_4_5 = pd.read_sql_query(query_4_5, connection)
answer_4_5

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\280175601.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_5 = pd.read_sql_query(query_4_5, connection)


count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

После более детального анализа вакансий получили:
- Наибольшим числом вакансии обладают регионы (Москва, Санкт-Петербург, Минск, Новосибирск, Алматы)
- В половине вакансиях указано хотя бы одно из двух полей с зарплатой (24073 вакансии)
- Средняя нижняя зарплатная граница вакансий равна 71065 рублей, средняя верхняя - 110537 рублей
- В наибольшем количестве вакансий указаны полный рабочий день и полная занятость
- В большинстве вакансий указан требуемый опыт работы от 1 года до 3 лет, меньше всего вакансий с опытом работы более 6 лет 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [175]:
query_5_1 = f'''select COUNT(*),
                    e.name
                from public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                GROUP BY e.name
                ORDER BY 1 DESC
            '''

In [176]:
answer_5_1 = pd.read_sql_query(query_5_1, connection)
answer_5_1.head(5)

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1971749194.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_1 = pd.read_sql_query(query_5_1, connection)


count           name
0   1933         Яндекс
1    491     Ростелеком
2    444       Тинькофф
3    428           СБЕР
4    331  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [177]:
query_5_2 = f'''select a.name,
                    COUNT(DISTINCT e.id),
                    COUNT(DISTINCT v.id)
                from public.areas a 
                LEFT JOIN public.employers e ON a.id = e.area
                LEFT JOIN public.vacancies v ON a.id = v.area_id
                WHERE v.id is null
                GROUP BY a.id
                ORDER BY 2 DESC
            '''

In [178]:
answer_5_2 = pd.read_sql_query(query_5_2, connection)
answer_5_2

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\729581154.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_2 = pd.read_sql_query(query_5_2, connection)


name  count  count
0                Россия    410      0
1             Казахстан    207      0
2    Московская область     75      0
3    Краснодарский край     19      0
4    Ростовская область     18      0
..                  ...    ...    ...
588             Скидель      0      0
589          Белоозерск      0      0
590           Орловский      0      0
591          Павловская      0      0
592             Таксимо      0      0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [179]:
query_5_3 = f'''select e.name,
                    COUNT(DISTINCT a.name) 
                from public.employers e
                JOIN public.vacancies v ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
                GROUP BY e.name
                ORDER BY 2 DESC
            '''

In [180]:
answer_5_3 = pd.read_sql_query(query_5_3, connection)
answer_5_3

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\3554497784.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_3 = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [181]:
query_5_4 = f''' 
                (select e.id   
                from public.employers e
                EXCEPT
                SELECT ei.employer_id
                FROM public.employers_industries ei )   
            '''

In [182]:
answer_5_4 = pd.read_sql_query(query_5_4, connection)
len(answer_5_4)

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1408135524.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_4 = pd.read_sql_query(query_5_4, connection)


8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [183]:
query_5_5 = f'''select e.name   
                from public.employers e
                JOIN public.employers_industries ei ON e.id = ei.employer_id
                GROUP BY e.name
                HAVING COUNT(ei.industry_id) = 4
                ORDER BY 1 ASC
                OFFSET 2 
                LIMIT 1
            '''

In [184]:
answer_5_5 = pd.read_sql_query(query_5_5, connection)
answer_5_5

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\3146325445.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_5 = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [185]:
query_5_6 = f'''select COUNT(*)   
                from public.employers e
                JOIN public.employers_industries ei ON e.id = ei.employer_id
                JOIN public.industries i ON i.id = ei.industry_id
                WHERE i.name = 'Разработка программного обеспечения'
               
            '''

In [186]:
answer_5_6 = pd.read_sql_query(query_5_6, connection)
answer_5_6

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\2173459197.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [187]:
#Код для получения городов миллионников
import requests 
from bs4 import BeautifulSoup 
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url) 
page = BeautifulSoup(response.text, 'html.parser') 
links = page.find_all('td')
lst_city = tuple([link.text for link in links[3:150:6]][0:16])


In [188]:
query_5_7 = f'''(select a.name,
                    COUNT(v.id)  
                from public.employers e
                JOIN public.vacancies v ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
                WHERE e.name = 'Яндекс' AND
                    a.name in {lst_city}
                GROUP BY a.name)
                UNION ALL
                (select 'total',
                    COUNT(*)  
                from public.employers e
                JOIN public.vacancies v ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
                WHERE e.name = 'Яндекс' AND
                    a.name in {lst_city})
                ORDER BY 2 ASC
            '''

In [189]:
answer_5_7 = pd.read_sql_query(query_5_7, connection)
answer_5_7

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\883904668.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_5_7 = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            total    485

***

После анализа работадателей:
- Наибольшее число вакансий имеют компании: Яндекс (1933), Ростелеком (491), Тинькофф, СБЕР, Газпромнефть
- Регионы, в которых нет вакансий и наибольшее число работодателей: Россия, Казахстан, что является ошибкой в данных (вакансии распределены по городам)
- Наибольшим охватом по регионам обладют компании: Яндекс, Ростелеком, Спецремонт (также Яндекс (181) и Ростелеком (152) обладают наибольшим числом вакансий)
- У 8419 работодателей не указана сфера их деятельности
- У 3553 работодателей указана сфера деятельности "Разработка программного обеспечения"
- У компании "Яндекс" 485 вакансий в городах-миллионниках (примерно четверть от общего числа вакансий компании "Яндекс")

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [190]:
query_6_1 = f'''select COUNT(*)
                from public.vacancies v
                WHERE LOWER(v.name) LIKE '%data%'
                    OR LOWER(v.name) LIKE '%данн%'
            '''

In [191]:
answer_6_1 = pd.read_sql_query(query_6_1, connection)
answer_6_1

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\401457554.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [192]:
query_6_2 = f'''select COUNT(*)
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
                    AND 
                    (LOWER(v.name) LIKE '%junior%'
                    OR LOWER(v.experience) LIKE '%нет опыта%'
                    OR LOWER(v.employment) LIKE '%стажировка%')
            '''

In [193]:
answer_6_2 = pd.read_sql_query(query_6_2, connection)
answer_6_2

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\4140246696.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [194]:
query_6_3 = f'''select COUNT(*)
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
                    AND
                    (LOWER(v.key_skills) LIKE '%sql%'
                    OR LOWER(v.key_skills) LIKE '%postgres%')
            '''

In [195]:
answer_6_3 = pd.read_sql_query(query_6_3, connection)
answer_6_3

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1417771425.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [196]:
query_6_4 = f'''select COUNT(*)
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
                    AND
                    LOWER(v.key_skills) LIKE '%python%'
            '''

In [197]:
answer_6_4 = pd.read_sql_query(query_6_4, connection)
answer_6_4

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1713433382.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [198]:
query_6_5 = f'''select ROUND(AVG(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, '\t', '')) + 1), 2)
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
            '''

In [199]:
answer_6_5 = pd.read_sql_query(query_6_5, connection)
answer_6_5

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\252717723.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [200]:
query_6_6 = f'''select v.experience,
                    ROUND(AVG(COALESCE((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)),0) average_salary
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
                    AND 
                    (v.salary_from IS NOT NULL
                    OR v.salary_to IS NOT NULL)
                GROUP BY v.experience
                ORDER BY 2 DESC  
            '''

In [201]:
answer_6_6 = pd.read_sql_query(query_6_6, connection)
answer_6_6

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\4219046839.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_6 = pd.read_sql_query(query_6_6, connection)


experience  average_salary
0       От 3 до 6 лет        243115.0
1  От 1 года до 3 лет        139675.0
2           Нет опыта         74643.0

***

После предметного анализа:
- 1771 вакансий имеют отношения к данным
- 51 вакансия с поиском начинающего специалиста по data science
- В 229 вакансиях для data science в ключевых навыках указаны SQL или postgres
- В 357 вакансиях для data science в ключевых навыках указан Python
- В среднем в вакансиях для data science указывают 6.41 ключевых навыков
- Заработная плата data science зависит от его опыта работы (средняя зп в вакансиях с опытом работы от 3 до 6 лет равна 243135, без опыта работы - 74643)

# Общий вывод по проекту

Общие выводы:
- Количество ваканский зависит от населения в городе, чем больше население, тем больше активных вакансий
- Наиболее распространенными компаниями по регионам являются "Яндекс" и "Ростелеком" (IT-компании), так как в IT-секторе есть возможность работать удаленно, что позволяет работать из большинства регионов и не требует открытия офиса для компании
- Общее количество ваканский DS равно 480
- В 357 вакансиях DS в ключевых навыках указан "Python", что примерно соответствует 3/4 от всех вакансий DS, следовательно знание языка программирования Python является наиболее распространенным навыком среди требований работодателя
- В 229 вакансиях DS в ключевых навыках указаны "SQL" или "postgres"что примерно соответствует половине от всех вакансий DS
- Средняя заработная плата DS зависит от опыта работы, чем больше опыт работы, тем выше заработная плата
- Наиболее распространенны вакансии DS с опытом работы от 1 до 3х лет и от 3х до 6 лет
- Низкое количество вакансий DS без опыта работы и с опытом работы от 6и лет означает, что чаще всего таких сотрудников ищут на профильных ресурсах

In [202]:
# Общее количество вакансий связанных с DS
query_6_7 = f'''select COUNT(*)
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
                    '''
answer_6_7 = pd.read_sql_query(query_6_7, connection)
answer_6_7

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\1730986655.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_7 = pd.read_sql_query(query_6_7, connection)


count
0    480

In [203]:
# Количество вакансий DS от требуемого опыта работы
query_6_8 = f'''select v.experience,
                    COUNT(*)
                from public.vacancies v
                WHERE (LOWER(v.name) LIKE '%data scientist%'
                    OR LOWER(v.name) LIKE '%data science%'
                    OR LOWER(v.name) LIKE '%исследователь данных%'
                    OR (v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%')
                    OR LOWER(v.name) LIKE '%machine learning%'
                    OR LOWER(v.name) LIKE '%машинн%обучен%')
                GROUP BY v.experience
                ORDER BY 2 DESC  
            '''
answer_6_8 = pd.read_sql_query(query_6_8, connection)
answer_6_8

C:\Users\dyaki\AppData\Local\Temp\ipykernel_2812\521321875.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_6_8 = pd.read_sql_query(query_6_8, connection)


experience  count
0  От 1 года до 3 лет    223
1       От 3 до 6 лет    195
2         Более 6 лет     34
3           Нет опыта     28

In [204]:
connection.close()